index.rst

```diff
@@ -48,9 +48,118 @@
+
+The Partial Solution
+====================
+
+A year or more ago we introduced a ``NO_SUDO`` flag environment variable
+that can be passed into the container, which bypasses all of the sudo
+steps to escalate privilege.
+
+This comes with one major caveat: it only works if there is a
+pre-provisioned user home directory.  This is fine at NCSA (which is
+where the rule was intended for use) since user home space is
+provisioned as part of account creation and NCSA forces all RSP users to
+have NCSA accounts; it is not fine in a GKE environment using GitHub as
+its authentication back-end.  We have recently proven that everything
+still works as expected (assuming a pre-provisioned filesystem) if the
+container is run with ``allowPrivilegeEscalation: False`` in its
+Kubernetes ``securityContext``.
+
+Recently, in order to allow loading the Multus CNI into containers in
+Rubin-managed Kubernetes environments, necessary for control of
+instruments on multicast networks with SAL, Josh Hoblitt introduced a
+container that must be run as a privileged initContainer inside the
+user's Lab pod before starting the Lab.  This change is already present
+in the mainline Nublado environment.
+
+A Fuller Solution
+=================
+
+Since we've worked out how to run privileged initContainers, all we
+would need to do is to move the provisioning steps into a second
+initContainer, and invoke it (or not) at user pod creation time.  We
+could either restrict its privilege still further by doing the same set
+of semi-privileged user plus sudoers rules we currently do, or just let
+the initContainer run as root, which would simplify the logic within
+considerably.
+
+Then, in all cases, we could invoke the Lab pod directly with the
+correct userid and set of groups, having provisioned those in ConfigMaps
+to mount into ``/etc/passwd`` and ``/etc/group`` (``NO_SUDO`` already
+does this), with the only piece that still requires privilege being the
+creation of user filesystems.
+
+This will be an initContainer that mounts our NFS share volume with
+``no_root_squash`` and effectively runs as root while doing so in order
+to create the user spaces.  That's a little scary, but no scarier than
+what we are doing today, particularly if we restrict the provisionator's
+actions with sudoers rules.
+
+Future Steps
+============
+
+At some point we very likely will to want to replace this mechanism with
+an actual JupyterHub-callable service that does the requested filesystem
+provisioning.  The good thing about the proposed model is that, when we
+do develop the service, then we just stop adding the privileged
+provisioning container to ``initContainers`` at user pod startup, and
+there's nothing else to do.  We cannot, with just the technique
+suggested here, set a cluster-wide ``PodSecurityPolicy``, since every
+user pod may have a privileged component.  However, the choice of
+implementation of the Multus CNI may have tied our hands with respect to
+this in any event.
```

Comment:
I don't see how Multus is relevant here since that's only the summit, correct?

Why not do this work now, rather than taking the temporary detour through init containers? I agree with Christine that the service approach seems all-around better, and we can trigger it on new account creation in COmanage down the road. It will also let us do other things that we may not want to expose to init containers running in the user's namespace, such as making GCE API calls to provision new Filestore volumes.

We can still use a `PodSecurityPolicy`; we just need a separate policy that allows containers to run as UID 0 and grant access to that policy only to the service that is responsible for initializing user directories.

Reply:
Multus is summit, TTS, and maybe base, at least.  Anywhere the RSP needs to control actual or simulated instruments.

I'm happy to do the service approach, but it's more effort.  I don't know if we care.  That's basically a call for Frossie to make, I think.